firewood/src/merkle/stream.rs

```diff
@@ -0,0 +1,528 @@
+// Copyright (C) 2023, Ava Labs, Inc. All rights reserved.
+// See the file LICENSE.md for licensing terms.
+
+use super::{node::Node, BranchNode, Merkle, NodeType, ObjRef};
+use crate::{
+    shale::{DiskAddress, ShaleStore},
+    v2::api,
+};
+use futures::Stream;
+use std::{ops::Not, task::Poll};
+
+pub(super) enum IteratorState<'a> {
+    /// Start iterating at the beginning of the trie,
+    /// returning the lowest key/value pair first
+    StartAtBeginning,
+    /// Start iterating at the specified key
+    StartAtKey(Vec<u8>),
+    /// Continue iterating after the given `next_node` and parents
+    Iterating {
+        visit_last: bool,
+        next_result: Option<(Vec<u8>, Vec<u8>)>,
```

Comment:
The state should not include the next key, only location information as to where we are.

As we discussed in-person, we should attempt to fix the existing stream iterator unless there's some reason why we can't.

Next steps are to add a unit test and debug the existing stream iterator code.